Bump SwarmKit to f74983e7c015a38a81c8642803a78b8322cf7eac

thaJeztah opened the pull request:
Full diff:

https://github.com/docker/swarmkit/compare/68a376dc30d8c4001767c39456b990dbd821371b...f74983e7c015a38a81c8642803a78b8322cf7eac


- Replace EC Private Key with PKCS#8 PEMs https://github.com/docker/swarmkit/pull/2246
- Fix IP overlap with empty EndpointSpec https://github.com/docker/swarmkit/pull/2505 (fixes https://github.com/moby/moby/issues/33795)
- Add support for Support SCTP port mapping (depends on changes in libnetwork) https://github.com/docker/swarmkit/pull/2298
- [orchestrator/updater] Do not reschedule tasks if only placement constraints change and are satisfied by the assigned node https://github.com/docker/swarmkit/pull/2496
- Ensure task reaper stopChan is closed no more than once https://github.com/docker/swarmkit/pull/2491
- [manager/dispatcher] Synchronization fixes (https://github.com/docker/swarmkit/pull/2495)

ping @anshulpundir @nishanttotla PTAL

thaJeztah committed:
Bump SwarmKit to f74983e7c015a38a81c8642803a78b8322cf7eac

- Replace EC Private Key with PKCS#8 PEMs
- Fix IP overlap with empty EndpointSpec
- Add support for Support SCTP port mapping (depends on changes in libnetwork)
- [orchestrator/updater] Do not reschedule tasks if only placement constraints change and are satisfied by the assigned node
- Ensure task reaper stopChan is closed no more than once
- [manager/dispatcher] Synchronization fixes

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
looks like several tests are failing

thaJeztah commented:
Failures are due to the changes in https://github.com/docker/swarmkit/pull/2246, so the tests may need some updating. Talked with @cyli on Slack, and she'll assist with updating the tests 👍

cyli commented:
@thaJeztah Tested this locally, and the swarm lock/unlock tests that failed all pass with this change:

```diff
diff --git a/integration-cli/docker_cli_swarm_test.go b/integration-cli/docker_cli_swarm_test.go
index 283f576f5..4a5ec9a56 100644
--- a/integration-cli/docker_cli_swarm_test.go
+++ b/integration-cli/docker_cli_swarm_test.go
@@ -4,7 +4,6 @@ package main

 import (
        "bytes"
-       "crypto/x509"
        "encoding/json"
        "encoding/pem"
        "fmt"
@@ -25,6 +24,7 @@ import (
        "github.com/docker/libnetwork/driverapi"
        "github.com/docker/libnetwork/ipamapi"
        remoteipam "github.com/docker/libnetwork/ipams/remote/api"
+       "github.com/docker/swarmkit/ca/keyutils"
        "github.com/go-check/check"
        "github.com/gotestyourself/gotestyourself/fs"
        "github.com/gotestyourself/gotestyourself/icmd"
@@ -1007,7 +1007,7 @@ func checkKeyIsEncrypted(d *daemon.Swarm) func(*check.C) (interface{}, check.Com
                        return fmt.Errorf("invalid PEM-encoded private key"), nil
                }

-               return x509.IsEncryptedPEMBlock(keyBlock), nil
+               return keyutils.IsEncryptedPEMBlock(keyBlock), nil
        }
 }
```

thaJeztah committed:
Update integration test to account for SwarmKit change

The PKCS8 changes updated the encryption on the keys so that the
`x509.IsEncryptedPEMBlock` may no longer return true because it cannot
parse the PEM block. The `keyutils` module in SwarmKit can tell whether
it is encrypted either way.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Applied the patch; let's see if CI goes green now 👍

vdemeester reviewed (approved):
LGTM 🐯

cpuguy83 commented on vendor/github.com/docker/swarmkit/manager/dispatcher/dispatcher.go:
```diff
@@ -532,6 +539,21 @@ func (d *Dispatcher) register(ctx context.Context, nodeID string, description *a
+	// Note that Stop() first does Dispatcher.ctx.cancel() followed by
+	// shutdownWait.Wait() to make sure new rpc's don't start before waiting
+	// for existing ones to finish.
+	d.shutdownWait.Add(1)
```
This looks suspect to me.
A waitgroup `Add()` is not safe to do async.
Ping @anshulpundir

anshulpundir replied:
sync.WaitGroup operations should be atomic. I think this is a false alarm. I will try to follow up with the community.

FYI: https://groups.google.com/forum/#!topic/golang-nuts/Qq_h0_M51YM

anshulpundir replied:
Actually it seems Add() needs to be synchronized with Wait(). Bummer.

I'll look at how to fix this.

cpuguy83 reviewed (commented):
Looks like this patch is pulling in a change from https://github.com/docker/swarmkit/pull/2495 that I'm a bit suspect of. The usage of the waitgroup looks racey on first glance.

anshulpundir reviewed (approved):
@nishanttotla and I discussed the data race failure in unit-tests and it should be OK to move forward. Because of how the waitgroup is used, the race should not have any undesired side-effects.

vdemeester reviewed (approved):
LGTM 🌵

thaJeztah commented:
Looks like https://github.com/docker/swarmkit/pull/2495 is being reverted in https://github.com/docker/swarmkit/pull/2518
